### ISSUE TITLE
pi 3 error after ap mode

### ISSUE THREAD

**Issue by emrekilic** [sabhiram/raspberry-wifi-conf#12]
I successfully installed raspberry-wifi-conf on RPi 3 with raspbian-jessie according to manual,  AP is started, I connected to AP, typed 192.168.44.1:88 to browser. I choose my ssid from list, typed my wifi password to dialog. When i press submit button, i got error from node side. Rpi3 didnt connect my network. Also after that i cant access to 192.168.44.1:88. AP mode is open. When i try to sudo npm start, i am getting this error everytime. Any help?

sudo npm start

> raspberry-wifi-conf@0.0.1 start /home/pi/Desktop/raspberry-wifi-conf
> node server.js

/home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:136
            info["hw_addr"].toLowerCase() == info["ap_addr"].toLowerCase() &&
                                                            ^

TypeError: Cannot read property 'toLowerCase' of undefined
    at _is_ap_enabled_sync (/home/pi/Desktop/raspberry-wifi-conf/app/wifi_manage                 r.js:136:61)
    at _is_wifi_enabled_sync (/home/pi/Desktop/raspberry-wifi-conf/app/wifi_mana                 ger.js:115:28)
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:126:35
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:90:20
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:251:                 17
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:154:                 25
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:248:                 21
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:612:                 34
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:76:17
    at ChildProcess.exithandler (child_process.js:204:7)

> raspberry-wifi-conf@0.0.1 start /home/pi/Desktop/raspberry-wifi-conf
> node server.js

/home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:136
            info["hw_addr"].toLowerCase() == info["ap_addr"].toLowerCase() &&
                                                            ^

TypeError: Cannot read property 'toLowerCase' of undefined
    at _is_ap_enabled_sync (/home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:136:61)
    at _is_wifi_enabled_sync (/home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:115:28)
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:126:35
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:90:20
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:251:17
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:154:25
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:248:21
    at /home/pi/Desktop/raspberry-wifi-conf/node_modules/async/lib/async.js:612:34
    at /home/pi/Desktop/raspberry-wifi-conf/app/wifi_manager.js:76:17
    at ChildProcess.exithandler (child_process.js:204:7)

npm ERR! Linux 4.1.18-v7+
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "start"
npm ERR! node v4.3.1
npm ERR! npm  v2.14.12
npm ERR! code ELIFECYCLE
npm ERR! raspberry-wifi-conf@0.0.1 start: `node server.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the raspberry-wifi-conf@0.0.1 start script 'node server.js'.
npm ERR! This is most likely a problem with the raspberry-wifi-conf package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node server.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls raspberry-wifi-conf
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/pi/Desktop/raspberry-wifi-conf/npm-debug.log

**Comment by sabhiram** [comment]
Hmm I am fairly busy w/ work and dont have immediate access to a RaspberryPi 3. Seems like the info structure in wifi manager is lacking the hw_addr. Let me see how that is figured out.

**Comment by sabhiram** [comment]
Can you paste the results of running this command on your Pi: 

`ifconfig wlan0`

**Comment by emrekilic** [comment]
ifconfig wlan0
wlan0     Link encap:Ethernet  HWaddr b8:27:eb:c1:0a:d5
          inet addr:192.168.44.1  Bcast:192.168.44.255  Mask:255.255.255.0
          inet6 addr: fe80::ba27:ebff:fec1:ad5/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2 errors:0 dropped:2 overruns:0 frame:0
          TX packets:97 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:125 (125.0 B)  TX bytes:19019 (18.5 KiB)

Thanks for helping me.

**Comment by Snyder1977** [comment]
It seems that I got stuck at the same point. I also have a pi 3.

Here is the error message and the output of `ifconfig wlan0`

`pi@raspberrypi:~/raspberry-wifi-conf $ sudo npm start

> raspberry-wifi-conf@0.0.1 start /home/pi/raspberry-wifi-conf
> node server.js


/home/pi/raspberry-wifi-conf/app/wifi_manager.js:136
            info["hw_addr"].toLowerCase() == info["ap_addr"].toLowerCase() &&
                                                             ^
TypeError: Cannot call method 'toLowerCase' of undefined
    at _is_ap_enabled_sync (/home/pi/raspberry-wifi-conf/app/wifi_manager.js:136:62)
    at _is_wifi_enabled_sync (/home/pi/raspberry-wifi-conf/app/wifi_manager.js:115:28)
    at /home/pi/raspberry-wifi-conf/app/wifi_manager.js:126:35
    at /home/pi/raspberry-wifi-conf/app/wifi_manager.js:90:20
    at /home/pi/raspberry-wifi-conf/node_modules/async/lib/async.js:251:17
    at /home/pi/raspberry-wifi-conf/node_modules/async/lib/async.js:154:25
    at /home/pi/raspberry-wifi-conf/node_modules/async/lib/async.js:248:21
    at /home/pi/raspberry-wifi-conf/node_modules/async/lib/async.js:612:34
    at /home/pi/raspberry-wifi-conf/app/wifi_manager.js:76:17
    at ChildProcess.exithandler (child_process.js:645:7)

npm ERR! raspberry-wifi-conf@0.0.1 start: `node server.js`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the raspberry-wifi-conf@0.0.1 start script.
npm ERR! This is most likely a problem with the raspberry-wifi-conf package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node server.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls raspberry-wifi-conf
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.1.19-v7+
npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "start"
npm ERR! cwd /home/pi/raspberry-wifi-conf
npm ERR! node -v v0.10.29
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/pi/raspberry-wifi-conf/npm-debug.log
npm ERR! not ok code 0
pi@raspberrypi:~/raspberry-wifi-conf $ ifconfig wlan0
wlan0     Link encap:Ethernet  HWaddr b8:27:eb:6c:ec:ac  
          inet addr:192.168.44.1  Bcast:192.168.44.255  Mask:255.255.255.0
          inet6 addr: fe80::ba27:ebff:fe6c:ecac/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2 errors:0 dropped:2 overruns:0 frame:0
          TX packets:390 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:125 (125.0 B)  TX bytes:79017 (77.1 KiB)`

Many thanks in advance. Any help would be very much appreciated.

**Comment by Snyder1977** [comment]
I have tried to get it running on an old pi model b. There I am getting the same error. So I guess it is not an issue that is only related to the new pi 3 model.

**Comment by jhermsen** [comment]
I had the same issue and found that iwconfig is not returning ssid and addr. So I changed the ap function in wifimanager to check if wlan IP address equals config ap address. Now script runs without error.
Now new problems remain.
I used rc.local to run script so after reboot it should work (also added logging), but for some reason script runs ok but ap is not available. If I than manually restart hostapd service it runs and ap is there.
If I then connect to the server I see the template without any ap's listed.

**Comment by Snyder1977** [comment]
Maybe it helps to add sudo in iwlist.js:

`exec("sudo iwlist scan", function(error, stdout, stderr)`

After adding "sudo" the ap´s were listed and I could enter the wifi credentials. I entered the credentials successfully and my pi was able to connect to my wifi network at home.  But then the next bugs turned up.  After the next reboot it seemed that the pi had forgotten the wifi credentials. It started in ap-mode and didn't connect properly to my wifi network. :-(

It would be great, if these issues were supported anyhow ....

**Comment by emrekilic** [comment]
@Snyder1977 I added sudo into iwlist.js but didnt change anything. When i run "sudo iwlist scan", i am getting this;
sudo iwlist scan
wlan0     Scan completed :
          Cell 01 - Address: 18:28:61:02:D9:5C
                    Channel:1
                    Frequency:2.412 GHz (Channel 1)
                    Quality=33/70  Signal level=-77 dBm
                    Encryption key:on
                    ESSID:"PALA"
                    Bit Rates:1 Mb/s; 2 Mb/s; 5.5 Mb/s; 11 Mb/s; 18 Mb/s
                              24 Mb/s; 36 Mb/s; 54 Mb/s
                    Bit Rates:6 Mb/s; 9 Mb/s; 12 Mb/s; 48 Mb/s
                    Mode:Master
                    Extra:tsf=0000000000000000
                    Extra: Last beacon: 40ms ago
                    IE: Unknown: 000450414C41 ...

wifi_manager.js
   var ifconfig_fields = {
        "hw_addr":         /HWaddr\s([^\s]+)/,
        "inet_addr":       /inet addr:([^\s]+)/,
    },  iwconfig_fields = {
        "ap_addr":         /Access Point:\s([^\s]+)/,
        "ap_ssid":         /ESSID:\"([^\"]+)\"/,
        "unassociated":    /(unassociated)\s+Nick/,
    }
iwlist wont have /Access Point:\s([^\s]+)/ attribute inside it. I changed it to  /Addess:\s([^\s]+)/, also didnt work. @jhermsen can you help me to change wifi_manager.js correctly. Thanks.

**Comment by Snyder1977** [comment]
Try to  install darland's fork. Maybe it works better due to some recent bug fixes...

**Comment by emrekilic** [comment]
@Snyder1977 After install darland's fork, error message disappeared. But now after restarting pi, it wont  connnect to wifi. Only connecting wifi in first time. After that Its always opening with Ap mode. Are there any other fork which solves this problems?

**Comment by Snyder1977** [comment]
I had the same problem first. I guess it has something to do with the boot sequence. It helped to add some sleeptime in rc.local:

`sudo nano /etc/rc.local`

`sleep 1m`

`cd /home/pi/raspberry-wifi-conf
sudo /usr/bin/node server.js < /dev/null &`

**Comment by emrekilic** [comment]
@Snyder1977 It worked in my first try. In first reboot pi connected  to my router and ap mode didnt enabled. After that i closed my router and restarted pi. Ap mode enabled because router closed. But i cant see rpi-config-ap from any of my devices. In screen i can see "ap enabled" but i cant see ssid. I opened my router and restarted pi again. I expect pi will connect to my router but didnt connect it. Ap mode enabled but i didnt see ssid again. Thanks.

**Comment by Snyder1977** [comment]
In AP mode I have to wait a couple of minutes to see the ssid.

But I think that there is still a bug in Darlands fork regarding the reconnection. If a previous known wifi network is turned off and then turned on again (e.g. night mode), then the pi does not reconnect to the previous known network. I am not able to fix this issue. Maybe this commit might help to fix it, please see lilianharris fork:

https://github.com/lillianharris/raspberry-wifi-conf/commit/57c9830a5cd122be6de2502a2d65802ee6d9d4b9

**Comment by emrekilic** [comment]
Thanks man. I waited at least 15 min. and didnt see ssid. Then i reboot pi. I saw ssid before rc.local runs. When server.js run inside rc.local, ap mode enabled again. After that ssid dissappearing. If i run server.js again, ap mode enabling at the same time. 

As i understand darland s fork wont support reconnection. If pi opens one time in ap mode, it never tries to connect router again. for solving this issue should i use all fork of lillianharris or only that commit?

**Comment by Snyder1977** [comment]
Did you set up the app as a service as described in the read me file?

`$sudo cp assets/init.d/raspberry-wifi-conf /etc/init.d/raspberry-wifi-conf`
`$sudo chmod +x /etc/init.d/raspberry-wifi-conf` 
`$sudo update-rc.d raspberry-wifi-conf defaults`

Regarding the reconnection I am cueless. I just saw lillian's commit. I tried to install lillian's fork but it didin't work on my pi 3. Maybe it helps to push lillians commit into darlands fork but I haven´t tried it.
